askama/src/lib.rs

```diff
@@ -76,23 +76,50 @@ pub use askama_escape::{Html, Text};
 pub trait Template {
     /// Helper method which allocates a new `String` and renders into it
     fn render(&self) -> Result<String> {
-        let mut buf = String::with_capacity(self.size_hint());
+        let mut buf = String::with_capacity(Self::SIZE_HINT);
         self.render_into(&mut buf)?;
         Ok(buf)
     }
     /// Renders the template to the given `writer` buffer
-    fn render_into(&self, writer: &mut dyn std::fmt::Write) -> Result<()>;
+    fn render_into<W: std::fmt::Write + ?Sized>(&self, writer: &mut W) -> Result<()>;
+
+    /// The template's extension, if provided
+    const EXTENSION: Option<&'static str>;
+
+    /// Provides an conservative estimate of the expanded length of the rendered template
+    const SIZE_HINT: usize;
+}
+
+pub trait DynTemplate {
+    fn dyn_render(&self) -> Result<String>;
+
+    fn dyn_render_into(&self, writer: &mut dyn std::fmt::Write) -> Result<()>;
+
     /// Helper function to inspect the template's extension
     fn extension(&self) -> Option<&'static str>;
+
     /// Provides an conservative estimate of the expanded length of the rendered template
     fn size_hint(&self) -> usize;
 }
 
-pub trait SizedTemplate {
+impl<T: Template> DynTemplate for T {
+    fn dyn_render(&self) -> Result<String> {
+        <Self as Template>::render(self)
+    }
+
+    fn dyn_render_into(&self, writer: &mut dyn std::fmt::Write) -> Result<()> {
+        <Self as Template>::render_into(self, writer)
+    }
+
     /// Helper function to inspect the template's extension
-    fn extension() -> Option<&'static str>;
+    fn extension(&self) -> Option<&'static str> {
```

Comment:
Should we mark these as `#[deprecated]`, and point to using `::EXTENSION` and `::SIZE_HINT`?

Reply:
They shouldn't be deprecated, because you can't access the associated consts from a `dyn DynTemplate`. Since `DynTemplate` is new, people using the no-argument form will get an error anyway and will hopefully be able to figure out they should use the associated consts.

Reply:
I added some more docs to both trait definitions, hopefully that will help.

Reply:
True